crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,248 @@
+                callback,
+            );
+        }
+
+        pub fn read_to_data_url<F>(self, blob: &impl BlobLike, callback: F)
+        where
+            F: FnOnce(Result<String, FileReadError>) + 'static,
+        {
+            self.read_as(
+                blob,
+                |js| js.as_string().unwrap_throw(),
+                web_sys::FileReader::read_as_data_url,
+                callback,
+            );
+        }
+
+        pub fn read_to_array_buffer<F>(self, blob: &impl BlobLike, callback: F)
+        where
+            F: FnOnce(Result<js_sys::ArrayBuffer, FileReadError>) + 'static,
+        {
+            self.read_as(
+                blob,
+                std::convert::Into::into,
+                web_sys::FileReader::read_as_array_buffer,
+                callback,
+            );
+        }
+
+        // fn on_progress<F>(&mut self, callback: F)
+        // where
+        //     F: FnMut(ProgressEvent) + 'static,
+        // {
+        //     unimplemented!()
+        // }
+
+        // fn on_load_start<F>(&mut self, callback: F)
+        // where
+        //     F: FnOnce(LoadStartEvent) + 'static,
+        // {
+        //     unimplemented!()
+        // }
+
+        fn read_as<T, ConvertFn, ReadFn, CallbackFn>(
+            self,
+            blob: &impl BlobLike,
+            convert: ConvertFn,
+            read: ReadFn,
+            callback: CallbackFn,
+        ) where
+            ConvertFn: Fn(wasm_bindgen::JsValue) -> T + 'static,
+            ReadFn: Fn(&web_sys::FileReader, &web_sys::Blob) -> Result<(), wasm_bindgen::JsValue>,
+            CallbackFn: FnOnce(Result<T, FileReadError>) + 'static,
+        {
+            let reader = self.inner.clone();
+            let closure = Closure::once(move || {
+                let result = reader
+                    .result()
+                    .map(convert)
+                    .map_err(std::convert::Into::into);
+                callback(result);
+            });
+            let function = closure.as_ref().dyn_ref().unwrap_throw();
+
+            self.inner.clone().set_onload(Some(&function));
+            read(&self.inner, blob.as_raw()).unwrap_throw();
+        }
+    }
+
+    impl std::ops::Drop for FileReader {
+        fn drop(&mut self) {
+            if self.inner.ready_state() < 2 {
+                self.inner.abort();
+            }
+        }
+    }
+
+}
+
+pub mod futures {
+    use super::FileReadError;
+    use crate::blob::BlobLike;
+    use futures::{sync::oneshot, Async, Future};
+    use wasm_bindgen::{closure::Closure, JsCast, UnwrapThrowExt};
+
+    #[derive(Debug)]
+    pub struct FileReader {
+        inner: web_sys::FileReader,
+    }
+
+    impl FileReader {
+        pub fn new() -> FileReader {
+            FileReader {
+                inner: web_sys::FileReader::new().unwrap_throw(),
+            }
+        }
+
+        pub fn read_as_string(self, blob: &impl BlobLike) -> ReadAsString {
```

Comment:
I think these should be top-level functions too.

Also, there's a lot of code duplication here. These should be using the `callbacks` functions and just wrapping them in Futures.

That means the `callbacks` functions need to return something which enables them to be dropped/cancelled.